scenes/game/ai_player.gd

```diff
@@ -328,40 +328,61 @@ func get_move_actions(game_logic : LocalGame, me : LocalGame.Player, opponent :
 				all_force_option_ids.append(card.id)
 			for card in me.gauge:
 				all_force_option_ids.append(card.id)
-			var combinations = []
-			generate_force_combinations(game_logic, me, all_force_option_ids, force_to_move_here, free_force_available, false, [], 0, combinations)
+			var combinations = generate_force_combinations(game_logic, me, all_force_option_ids, force_to_move_here, free_force_available)
 			for combo_result in combinations:
 				var combo = combo_result[0]
 				var used_free_force = combo_result[1]
 				possible_move_actions.append(MoveAction.new(i, combo, used_free_force))
 	return possible_move_actions
 
-func generate_force_combinations(game_logic : LocalGame, me : LocalGame.Player, cards, force_target, free_force_available, used_free_force, current_combination, current_index, combinations):
+## Given a list of card IDs `cards`, return all combinations of card IDs that can be spent to generate
+## `force_target` Force, with up to `free_force_available` units worth of slack. The actual return value
+## is a list of pairs [combination, free_force_used], where the latter is true iff at least one point of
+## the "slack" is necessary for the combination.
+func generate_force_combinations(game_logic : LocalGame, me : LocalGame.Player, cards, force_target, free_force_available):
 	var current_force = me.force_cost_reduction
 	var card_db = game_logic.get_card_database()
-	for card_id in current_combination:
-		current_force += card_db.get_card_force_value(card_id)
-	if current_force >= force_target:
-		combinations.append([current_combination.duplicate(), used_free_force])
-		return
-
-	if free_force_available > 0:
-		generate_force_combinations(game_logic, me, cards, force_target - free_force_available, 0, true, current_combination, current_index, combinations)
-
-	for i in range(current_index, cards.size()):
-		current_combination.append(cards[i])
-		generate_force_combinations(game_logic, me, cards, force_target, 0, used_free_force, current_combination, i + 1, combinations)
-		current_combination.pop_back()
-
-func generate_card_count_combinations(cards, hand_size, current_combination, current_index, combinations):
-	if current_combination.size() == hand_size:
-		combinations.append(current_combination.duplicate())
-		return
-
-	for i in range(current_index, cards.size()):
-		current_combination.append(cards[i])
-		generate_card_count_combinations(cards, hand_size, current_combination, i + 1, combinations)
-		current_combination.pop_back()
+	if current_force == force_target:
+		return [[[], false]]
+
+	var candidates = [[current_force]]  # Each entry in result is a list whose car is a force count and whose cdr is a list of cards
+	for card_id in cards:
```

Comment:
How does this not just fill up the candidates with the first X cards for your force target of X?
I'm not quite following this algorithm.

Reply:
Oh, it only appends more candidates, it doesn't add to existing candidates.
I think I get it now.

Reply:
Yeah; we add candidates because when we're thinking about a particular card X we want all subsets that can contain X, but we still want to have all the subsets that don't contain X in our list because we might (duplicate and) add other stuff to them later.